### PR TITLE
[None][feat] Cudagraph updates for helix parallelism

### PR DESCRIPTION
**Pull request by brb-nv**
## Description

Currently, cuda graph functionality for helix parallelism is broken resulting in sporadically poor accuracy in CI. This MR makes the following changes:
- Use static buffers for `helix_position_offsets` and `helix_is_inactive_rank` whether cudagraph is enabled or not.
- Unify where these tensors are assigned and used. Current assignment and usage is non-uniform and confusing.
- Update e2e tests with cudagraph parameterization.

## Test Coverage
```
# No cuda graph.
$ pytest tests/integration/defs/accuracy/test_disaggregated_serving.py::TestDeepSeekV3Lite::test_auto_dtype_with_helix[cudagraph:none] -s -v

# Without padding.
$ pytest tests/integration/defs/accuracy/test_disaggregated_serving.py::TestDeepSeekV3Lite::test_auto_dtype_with_helix[cudagraph:without_padding] -s -v

# With padding.
$ pytest tests/integration/defs/accuracy/test_disaggregated_serving.py::TestDeepSeekV3Lite::test_auto_dtype_with_helix[cudagraph:with_padding] -s -v
```

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

## Release Notes

* **New Features**
  * Added Helix parallelism support for distributed attention processing with improved parameter management and state tracking.
  
* **Refactor**
  * Reorganized helix parameter handling through centralized metadata flow for enhanced compatibility with CUDA graphs and distributed configurations.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 1]
```diff
@@ -1689,23 +1692,35 @@ def _prepare_tp_inputs(
...
-                    # Warmup doesn't have `total_input_len_cp` set because merge_helix_requests is not called.
-                    if not self.is_warmup and not request.is_cuda_graph_dummy:
-                        position_id = request.total_input_len_cp + request.py_decoding_iter - 1
+                    #####################################################################################
+                    assert not self.is_warmup, "Warmup is not called for helix parallelism."
+                    helix_is_inactive_rank_all_ranks = self.dist.cp_allgather(
+                        request.py_helix_is_inactive_rank)
+                    helix_is_inactive_rank_all_ranks = torch.tensor(
+                        helix_is_inactive_rank_all_ranks, dtype=torch.long)
+                    assert torch.sum(
```
Comment (brb-nv) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: @brb-nv , remove this expensive check.

**Review by thorjohnsen — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/modules/attention.py [hunk 2]
```diff
@@ -733,7 +733,7 @@ def __init__(
             dtype (torch.dtype): The data type.
             dense_bias (bool): Whether to use bias in the output projection layer.
             config (ModelConfig): The model configuration.
-            enable_unit_test (bool): Whether to enable unit test.
+            enable_helix_test (bool): Whether to enable unit test.
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/attention.py]: "Whether to enable unit test" -> "Whether to enable helix test"

File: tensorrt_llm/_torch/modules/attention.py [hunk 3]
```diff
@@ -1102,14 +1102,12 @@ def _attn_forward_gen(self, attn_backend: AttentionBackend, q: torch.Tensor,
...
-                attn_metadata,
-                softmax_stats_tensor=softmax_stats,
-                helix_position_offsets=position_ids,
-                **kwargs)
+            partial_o = attn_backend.forward(q,
+                                             k,
+                                             v,
+                                             attn_metadata,
+                                             softmax_stats_tensor=softmax_stats,
+                                             **kwargs)
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/attention.py]: Can add a "," to keep format style unchanged

File: tensorrt_llm/_torch/modules/attention.py [hunk 4]
```diff
@@ -1819,18 +1804,9 @@ def forward_absorption_generation(
...
-                    cu_kv_seqlens,
-                    fmha_scheduler_counter,
-                    mla_bmm1_scale,
-                    mla_bmm2_scale,
-                    quant_q_buffer,
-                    helix_position_offsets=helix_position_offsets,
-                    helix_is_inactive_rank=helix_is_inactive_rank),
+                    fused_q, q_pe, latent_cache, attn_metadata, cu_q_seqlens,
+                    cu_kv_seqlens, fmha_scheduler_counter, mla_bmm1_scale,
+                    mla_bmm2_scale, quant_q_buffer),
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/attention.py]: Can add a "," after quant_q_buffer to keep style unchanged

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 5]
```diff
@@ -1689,23 +1692,35 @@ def _prepare_tp_inputs(
 
                 position_id = past_seen_token_num
                 if self.mapping.has_cp_helix():
-                    # Warmup doesn't have `total_input_len_cp` set because merge_helix_requests is not called.
-                    if not self.is_warmup and not request.is_cuda_graph_dummy:
-                        position_id = request.total_input_len_cp + request.py_decoding_iter - 1
+                    #####################################################################################
```
Comment (yuxianq) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: Remove this line

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 6]
```diff
@@ -1689,23 +1692,35 @@ def _prepare_tp_inputs(
...
+                    assert not self.is_warmup, "Warmup is not called for helix parallelism."
+                    helix_is_inactive_rank_all_ranks = self.dist.cp_allgather(
+                        request.py_helix_is_inactive_rank)
+                    helix_is_inactive_rank_all_ranks = torch.tensor(
+                        helix_is_inactive_rank_all_ranks, dtype=torch.long)
+                    assert torch.sum(
+                        helix_is_inactive_rank_all_ranks
+                    ) == self.mapping.cp_size - 1, "There should be only one active rank in helix parallelism."
+                    #####################################################################################
```
Comment (yuxianq) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: Remove this line

Comment (brb-nv) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Done. https://github.com/NVIDIA/TensorRT-LLM/pull/10141/commits/685bc1300bad9280aebec69aa1420c0f658b16ba

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 7]
```diff
@@ -570,12 +570,13 @@ def warmup(self, resource_manager: ResourceManager) -> None:
...
 
-        if self.mapping.cp_size > 1:
-            cp_type = self.mapping.cp_config.get("cp_type", None)
-            logger.info(
-                f"[ModelEngine::warmup] Skipping warmup for cp_type: {None if cp_type is None else cp_type.name}."
-            )
-            return
+        cp_type = self.mapping.cp_config.get('cp_type', None)
+        if cp_type is not None:
+            if cp_type in [CpType.ULYSSES, CpType.STAR, CpType.HELIX]:
```
[the remark this answers is not in the document]
Comment (brb-nv) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Great catch. Fixed. https://github.com/NVIDIA/TensorRT-LLM/pull/10141/commits/685bc1300bad9280aebec69aa1420c0f658b16ba

File: tensorrt_llm/_torch/modules/attention.py (hunk 3 above)
Comment (brb-nv) [reply tensorrt_llm/_torch/modules/attention.py]: Looks a lot better. https://github.com/NVIDIA/TensorRT-LLM/pull/10141/commits/c84a95dde5e1ba2d6f347a2d2e0bdfbff9c754bd

**Review by yuxianq — APPROVED** [review]
